### ISSUE TITLE
Proposal: Allow for full control over the styling of NavigationViewItemSeparator

### ISSUE THREAD

**Issue by joshkoon** [microsoft/microsoft-ui-xaml#2139]
# Proposal: Allow for full control over the styling of NavigationViewItemSeparator

## Summary
@joshkoon:
NavigationViewItemSeparator supports the Style property, but ignores many of the setters. Proposal is to update the control to allow apps to have more control over the appearance.

--------------------------------------------------------------------------------------------------------------
@anawishnoff:
ThemeResources will be added for various styling properties of NavigationViewItemSeparator to allow developers to have a lightweight styling experience and customize the look of their separators without having to edit a control template.

**Solution proposed by @Felix-Dev.**

## Rationale

@joshkoon:
* Inflexibility of control appearance forces developers to retemplate for trivial changes, such as margins.

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to customize the appearance of the Separator | Must |
| ThemeResources will be created for various styling properties of NavigationViewItemSeparator | Must |
| NavigationViewItemSeparator will use new ThemeResources within its ControlTemplate | Must |

## Important Notes

This proposal aims to create ThemeResources in the [NavigationView Resource Dictionary](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/NavigationView/NavigationView_rs1_themeresources.xaml):

Margin:
```xml
<Thickness x:Key="TopNavigationViewItemSeparatorMargin">10,0</Thickness>
```
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">16,10</Thickness>
```
```xml
<Thickness x:Key="NavigationViewCompactItemSeparatorMargin">16,10</Thickness>
```

Foreground:
```xml
<StaticResource x:Key="NavigationViewItemSeparatorForeground" ResourceKey="SystemControlForegroundBaseLowBrush" />
```
```xml
<StaticResource x:Key="TopNavigationViewItemSeparatorForeground" ResourceKey="SystemControlForegroundBaseLowBrush" />
```

Height for left-docked NavigationView:
```xml
<x:Double x:Key="NavigationViewItemSeparatorHeight">1</x:Double>
```

Width for top-docked NavigationView:
```xml
<x:Double x:Key="TopNavigationViewItemSeparatorWidth">1</x:Double>
```

**Comment by StephenLPeters** [comment]
@ojhad  fyi

**Comment by anawishnoff** [comment]
This is the control template for NavigationViewItemSeparator:

```xml
 <Style TargetType="local:NavigationViewItemSeparator">
        <Setter Property="IsEnabled" Value="False" />
        <Setter Property="IsTabStop" Value="False" />
        <Setter Property="MinHeight" Value="0" />
        <Setter Property="AutomationProperties.AccessibilityView" Value="Raw" />
        <Setter Property="Template">
            <Setter.Value>
                <ControlTemplate>
                    <Grid x:Name="NavigationViewItemSeparatorRootGrid">
                        <VisualStateManager.VisualStateGroups>
                            <VisualStateGroup x:Name="NavigationSeparatorLineStates">
                                <VisualState x:Name="HorizontalLine" />
                                <VisualState x:Name="VerticalLine">
                                    <VisualState.Setters>
                                        <Setter Target="SeparatorLine.Height" Value="20" />
                                        <Setter Target="SeparatorLine.Width" Value="1" />
                                        <Setter Target="SeparatorLine.Margin" Value="10,0" />
                                        <Setter Target="SeparatorLine.VerticalAlignment" Value="Center" />
                                    </VisualState.Setters>
                                </VisualState>
                            </VisualStateGroup>
                        </VisualStateManager.VisualStateGroups>
                        <Rectangle 
                            x:Name="SeparatorLine"
                            Height="1"
                            Margin="16,10"
                            Fill="{ThemeResource SystemControlForegroundBaseLowBrush}" />
                    </Grid>
                </ControlTemplate>
            </Setter.Value>
        </Setter>
    </Style>
```

@ojhad 
Could this be fixed by using TemplateBindings in the setter property values for Height, Width, Margin, and VerticalAlignment (or whichever ones we choose to make customizable)?

@joshkoon 
Are there any other properties that you were specifically interested in changing?

**Comment by Felix-Dev** [comment]
@anawishnoff I'm currently looking at issue https://github.com/microsoft/microsoft-ui-xaml/issues/1961 locally here and would give it a shot if theme resources should be introduced. As such, I would add at least a theme resource for the NavigationViewItemSeparator specifying its margin. The rest of the NavigationViewItemSeparator properties I would leave untouched in that PR. Any thoughts?

**Comment by anawishnoff** [comment]
@Felix-Dev - how would this ThemeResource be used? Would it serve as the default for the NavigationViewItemSeparator Margin property as defined above in this snippet?
```xaml
<Rectangle 
    x:Name="SeparatorLine"
    Height="1"
    Margin="16,10"
    Fill="{ThemeResource SystemControlForegroundBaseLowBrush}" />
```

Looking more deeply into #1961 and your comment, I think adding the ThemeResources is a good idea. Just want to make sure it won't conflict with anything that we implement to solve this issue/clarify what issue the ThemeResources would be solving.

**Comment by Felix-Dev** [comment]
Yes, these theme resources would act as the default margins for the NavigationViewItemSeparator, customizable by the developer/designer. As such, the updated style would look like this:

```xml
<Style TargetType="local:NavigationViewItemSeparator">
    <Setter Property="IsEnabled" Value="False" />
    <Setter Property="IsTabStop" Value="False" />
    <Setter Property="MinHeight" Value="0" />
    <Setter Property="AutomationProperties.AccessibilityView" Value="Raw" />
    <Setter Property="Template">
        <Setter.Value>
            <ControlTemplate>
                <Grid x:Name="NavigationViewItemSeparatorRootGrid">
                    <VisualStateManager.VisualStateGroups>
                        <VisualStateGroup x:Name="NavigationSeparatorLineStates">
                            <VisualState x:Name="HorizontalLine" />
                            <VisualState x:Name="VerticalLine">
                                <VisualState.Setters>
                                    <Setter Target="SeparatorLine.Height" Value="20" />
                                    <Setter Target="SeparatorLine.Width" Value="1" />
                                    <Setter Target="SeparatorLine.Margin" Value="{StaticResource TopNavigationViewItemSeparatorMargin}" />
                                    <Setter Target="SeparatorLine.VerticalAlignment" Value="Center" />
                                </VisualState.Setters>
                            </VisualState>
                        </VisualStateGroup>
                    </VisualStateManager.VisualStateGroups>
                    <Rectangle 
                        x:Name="SeparatorLine"
                        Height="1"
                        Margin="{StaticResource NavigationViewItemSeparatorMargin}"
                        Fill="{ThemeResource SystemControlForegroundBaseLowBrush}" />
                </Grid>
            </ControlTemplate>
        </Setter.Value>
    </Setter>
</Style>
```

The theme resources would be defined in [NavigationView_rs1_themeresources.xaml](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/NavigationView/NavigationView_rs1_themeresources.xaml) presumably.

**Comment by anawishnoff** [comment]
@Felix-Dev Thanks for clarifying! Make sense now. So with these ThemeResources in place, is the idea that you could edit them directly rather than having to edit the control template/use a setter property to edit the styling on a NavigationViewItemSeparator?

That would conflict a little bit with this issue, in my opinion. I think we'd want to make sure that any customizable property of NavigationViewItemSeparator can be customized in the same manner. If only margin had a ThemeResource defined, wouldn't that throw a wrench in things?

**Comment by Felix-Dev** [comment]
> So with these ThemeResources in place, is the idea that you could edit them directly rather than having to edit the control template/use a setter property to edit the styling on a NavigationViewItemSeparator?

Absolutely! Here's an example how the to-be-introduced TopNavigationViewItemSeparatorMargin theme resource could be used:
```xml
<muxc:NavigationView PaneDisplayMode="Top">
    <muxc:NavigationView.Resources>
        <Thickness x:Key="TopNavigationViewItemSeparatorMargin">2,0,2,0</Thickness>
    </muxc:NavigationView.Resources>
</muxc:NavigationView>
```

> That would conflict a little bit with this issue, in my opinion. I think we'd want to make sure that any customizable property of NavigationViewItemSeparator can be customized in the same manner. If only margin had a ThemeResource defined, wouldn't that throw a wrench in things?

In case I would get the go-ahead to try working on improving lightweight styling for the NavigationView, I would gladly introduce other theme resources excluded from my initial thought (which was specifically focused on making the margins of NavigationView pane items easily customizable (as asked for in issue #1961). This issue here could warrant adding additional resources outside of those specific margin resources.

**Comment by Felix-Dev** [comment]
Ok, I made some quick tests with the width and height of a NavigationViewItemSeparator (width in display mode top, height in left mode) and the results seem presentable enough:
![image](https://user-images.githubusercontent.com/1398851/77699843-6b679100-6fb3-11ea-9f9a-6817ad8e9d69.png)
(That's a height of 2px for the separator, which originally has a height of 1px). And what's a nice looking UI is subject to differ for any two folks anyway. The point of lightweight styling is that the UI can easily be modified without having to create a new Style (let alone a hard-copy of a Style!). As such, I now think we should also add Width and Height controlling theme resources for the NavigationViewItem separator:
```xml
<x:Double x:Key="NavigationViewItemSeparatorHeight">1</x:Double>
<x:Double x:Key="TopNavigationViewItemSeparatorWidth">1</x:Double>
```

**Comment by YuliKl** [comment]
Overall, I think this is a great approach.

> As for other NavigationViewItemSeparator properties...should we also make Width (in top mode) or Height in left mode for the separator stylable via lightweight theme resources (@joshkoon to comment)? Is there demand to customize these values too?

I wonder if specifically for height and width, TemplateBinding to the properties would be more intuitive than using lightweight styling. But I'd love for @kikisaints, the mistress of all things light weight styling, to weigh in.

**Comment by YuliKl** [comment]
Side comment - I'm seeing something really strange happening with the padding of separators in the latest version of the Controls Gallery.
![image](https://user-images.githubusercontent.com/13142665/77705410-d2b61e80-6f7c-11ea-9ba8-1ac5d0c2f45d.png)
Maybe new padding theme resources will fix this bug 🙂

**Comment by Felix-Dev** [comment]
@YuliKl I was thinking that if we already introduce lightweight styling theme resources like Foreground and Margin for the separator, we could also just add those two Width and Height theme resources as well. That way, the developer would not have to create a new style just because they wanted to change the "thickness" *of all the separators* used in a NavigationView. Especially if all the other properties of the separator they might want to change are available as theme resources.

**Comment by anawishnoff** [comment]
@Felix-Dev I'm excited about this approach and agree with your point above regarding the Width and Height ThemeResources - best those be included as well to avoid confusion. Is this something you were looking to tackle in a PR? 😊

**Comment by Felix-Dev** [comment]
@anawishnoff Yes, I would like to address this issue in a PR 🙂 Let me know when you think we are ready to start working it!

**Comment by anawishnoff** [comment]
@Felix-Dev Awesome! As we aren't necessarily changing an API, I don't think this requires a full spec. @ranjeshj correct me if I'm wrong? If I'm correct, a PR can be started whenever you have capacity.

**Comment by Felix-Dev** [comment]
@anawishnoff Working on the PR, I have made two additional observations for now:

#### Observation 1:
Introducing the theme resource
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">16,10</Thickness>
```
can cause issues for a navigation view in `left compact `mode when this resource is set to a margin using larger values for the left and right margin. Take for example the setting 
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">32,10</Thickness>
```
which causes the separator to not be visible any longer in `left compact` mode:
![image](https://user-images.githubusercontent.com/1398851/77958964-28206180-72d6-11ea-8a4a-427d53c8ad5a.png)

With the original values `16,10` the separator is clearly visible (in red):
![image](https://user-images.githubusercontent.com/1398851/77959103-5d2cb400-72d6-11ea-9b67-82c19d863901.png)

As such, I propose the following solution here: 
Currently, the `NavigationViewItemSeparator` has only two visual states: `HorizontalLine` and `VerticalLine`. We could introduce a *third* visual state named `HorizontalLineCompact` which would set a specific margin value to the NavigationViewItemSeparator: 
```xml
<Thickness x:Key="NavigationViewCompactItemSeparatorMargin">16,10</Thickness>
```
As such, the updated style would now look like this (excluding the irrelevant parts):
```xml
<Style TargetType="local:NavigationViewItemSeparator">
    <!-- ... -->
    <Setter Property="Template">
        <Setter.Value>
            <ControlTemplate>
                <Grid x:Name="NavigationViewItemSeparatorRootGrid">
                    <VisualStateManager.VisualStateGroups>
                        <VisualStateGroup x:Name="NavigationSeparatorLineStates">
                            <VisualState x:Name="HorizontalLine" />
                            <VisualState x:Name="HorizontalLineCompact">
                                <VisualState.Setters>
                                    <Setter Target="SeparatorLine.Margin" Value="{ThemeResource NavigationViewCompactItemSeparatorMargin}" />
                                </VisualState.Setters>
                            </VisualState>
                            <VisualState x:Name="VerticalLine">
                                <VisualState.Setters>
                                    <!-- ... -->
                                    <Setter Target="SeparatorLine.Margin" Value="{ThemeResource TopNavigationViewItemSeparatorMargin}" />
                                </VisualState.Setters>
                            </VisualState>
                        </VisualStateGroup>
                    </VisualStateManager.VisualStateGroups>
                    <Rectangle
                        x:Name="SeparatorLine"
                        Margin="{ThemeResource NavigationViewItemSeparatorMargin}"/>
                </Grid>
            </ControlTemplate>
        </Setter.Value>
    </Setter>
</Style>
```

#### Observation 2:
We can set the foreground of a `NavigationViewItem` by setting its *Foreground* property. This currently does not work for a NavigationViewItemSeparator (settings its foreground property has no effect). Should we thus "enable" the Foreground property on it too? 
If yes, we probably need to think some more about the potential implementation because if we use TemplateBinding in the current style, the user-provided foreground color would be overriden as soon as the separator is used in navigation view display mode "top" (because we are overriding the foreground color in a visual state setter with the specific `TopNavigationViewItemSeparatorForeground` resource). As such, supporting setting the Foreground property and having it work for all navigation view display modes might not actually be straight-forward.

Thoughts?

**Comment by anawishnoff** [comment]
@Felix-Dev  Just to clarify about your first point: If we made a `NavigationViewCompactItemSeparatorMargin` resource, one would have to define their desired item separator styling for non-compact mode with `NavigationViewItemSeparatorMargin` and separately define their desired item separator styling for compact mode with `NavigationViewCompactItemSeparatorMargin`? 

I'm trying to weigh the pros and cons of either having the separator's margin be un-editable in left compact mode, or adding another resource that could make the styling process less intuitive, i.e. "why isn't the margin changing in non-compact mode but not in compact mode?"

Do we understand why the separator is disappearing in compact mode? I was under the impression that the margin values defined were for top and bottom, which shouldn't be pushing the separator out of the frame.

**Comment by Felix-Dev** [comment]
@anawishnoff Yes, my proposed solution would mean we would now have *two* theme resources to set the margin of the separator in keft mode. One resource to be applied when the navigation view pane is expanded and one resource to be used when the pane is closed (compact mode). This additional theme resource in compact mode is required because of the following reason:

Each theme resource is a `Thickness` value which enables to set the margin to the left, top, right and bottom. In other words, a resource like 
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">16,10</Thickness>
```
could also be written equivalently as
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">16,10,16,10</Thickness>
```
with Left = Right =16 & Top = Bottom = 10

In top mode, we only need one theme resource because there is no compact mode. In left mode, however, if we decide to add a theme resource for the margin, the user might change our default `16,10` margin to a different margin, like 
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">32,10,32,10</Thickness>
```
This causes the separator to not show up any longer in the left compact mode of the Navigation view (with the navigation pane in a closed state), because there is no space left for it any longer. The first 32 pixel from the left border to the center and the first 32 pixel from the right border to the center are now "margin pixels", that is 64 pixel of the navigation view compact mode (on the horizontal axis) are no longer available to our actual separator line. And since the navigation view column doesn't have more than 64 pixel in its compact state with the pane closed, there are no pixels left for the actual separator. To aid in understanding this, please look at the following resource values and their corresponding visual representation:

```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">16,10,0,10</Thickness>
```
![image](https://user-images.githubusercontent.com/1398851/78072203-00e19700-739f-11ea-9ecd-d4f5aa4b2efc.png)

Now we are pushing it further to the right by increasing its left margin:
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">32,10,0,10</Thickness>
```
![image](https://user-images.githubusercontent.com/1398851/78072263-18208480-739f-11ea-9060-366d44dbb4ef.png)

and finally we are also adding some margin to the right now:
```xml
<Thickness x:Key="NavigationViewItemSeparatorMargin">32,10,32,10</Thickness>
```
![image](https://user-images.githubusercontent.com/1398851/78072421-66ce1e80-739f-11ea-9978-eae1c3462406.png)

The whole images were probably overkill right now as I'm sure you got the problem the moment you read the margins are applied to to not only top and bottom, but also left and right 😅

That said, this should illustrate the problem why we probably need *two* margin resources for the left mode of the navigation view *if* we consider users could also change the left and right margin of the separator and not just the top and bottom margins. Left mode has both an expanded mode (pane opened) and a compact mode (pane closed) and the two margin resources for the separator would reflect that.

Of course, we could decide, as you said, to not allow separator margins to be changeable if in closed compact mode. In that case, I would still have the third internal visual state, but this time, there would be no additional theme resource `NavigationViewCompactItemSeparatorMargin`. I am wondering though...suppose the user did provide their own values for the `NavigationViewItemSeparatorMargin` theme resource (so they modified the separator margin in left mode) this could introduce unwelcome "visual jumps" for the separator when switching from compact mode to expanded mode and vice-versa (as now different margins for top and bottom could be used, causing the separator to visually move up or down).

**Comment by anawishnoff** [comment]
@Felix-Dev Thanks so much for the detailed explanation! Definitely makes sense, and I agree with your point about having weird visual jumps if we allowed editing in expanded mode but not compact mode.

I think adding another resource for `NavigationViewCompactItemSeparatorMargin` is a good and necessary approach. I'll update the proposal to include this.

As for the Foreground issue: I don't think this PR should be blocked on needing to implement foreground resources for NavigationViewItemSeparator. Changing the foreground color specifically was not requested, and if it's going to require non-trivial API changes it's best that it becomes its own issue, in my opinion. I'd go ahead and focus on the other resources if you agree.

**Comment by Felix-Dev** [comment]
> As for the Foreground issue: I don't think this PR should be blocked on needing to implement foreground resources for NavigationViewItemSeparator.

That's good to know :) And not "TemplateBinding" the Foreground property doesn't mean the foreground won't be customizable. The PR would still add *theme resources* the developer/designer can use to set a custom foreground color for the separator in either display mode, if so required. Only the Foreground property approach can't be used but we can address this in a future commit if there is user demand adding this support too, just as you suggested.

**Comment by anawishnoff** [comment]
@Felix-Dev Thanks again for clarifying - sounds good!
